CI red on master (fac47cd): ruff-format drift + Windows file-lock in test_forecast

## Summary

Two CI checks fail on `master` (`fac47cd`) independently of any PR — confirmed on the master commit's own check-runs, so every open PR inherits these reds:

```
$ gh api repos/davidusb-geek/emhass/commits/fac47cd/check-runs \
    --jq '.check_runs[] | select(.conclusion=="failure") | .name'
build (ubuntu-latest)
build (windows-latest)
```

## 1. `ruff format --check` drift (build → "Check formatting" step)

The `uvx ruff format --check --diff` step fails: two files on master are not ruff-formatted.

```
2 files would be reformatted, 47 files already formatted
```

Affected files:
- `src/emhass/forecast.py` (e.g. the stale Open-Meteo branch around line 421 — a call wrapped across 3 lines that ruff would collapse to one)
- `tests/test_forecast.py` (e.g. a `w_forecast_cache_path = ...` assignment around line 1554 that ruff would wrap in parens)

**Fix:** `uvx ruff format src/emhass/forecast.py tests/test_forecast.py` and commit.

## 2. Windows file-lock in the stale Open-Meteo cache test

`build (windows-latest)` fails the test run:

```
FAILED tests/test_forecast.py::TestForecast::test_get_cached_forecast_data_stale_open_meteo_deletes_cache
  - PermissionError: [WinError 32] The process cannot access the file because it is being
    used by another process: 'D:\a\emhass\emhass\data\weather_forecast_data_stale_test.pkl'
```

On Windows a file can't be unlinked/reopened while a handle is still open. The stale-cache removal path in `forecast.py` appears to delete/reopen the `.pkl` while its handle is still held — fine on POSIX, fails on Windows. `build (macos-latest)` is also red (~45 min, runner timeout in the same suite region).

**Fix direction:** ensure the cache file handle is closed (e.g. `with open(...)` context manager) before `unlink()`/re-read in the stale-cache branch.

## Repro

```
gh api repos/davidusb-geek/emhass/commits/fac47cd/check-runs \
  --jq '.check_runs[] | select(.conclusion=="failure") | .name'
```

Filed to keep the signal visible — happy to send a PR for the `ruff format` part if useful; the Windows handle fix touches `forecast.py` runtime behaviour and is better owned by a maintainer.

--- a/src/emhass/forecast.py
+++ b/src/emhass/forecast.py
@@ -1844,59 +1844,61 @@ class Forecast:
         :rtype: pd.DataFrame | None
 
         """
+        # Read then close the file BEFORE any os.remove below: on Windows a file
+        # cannot be unlinked while a handle is still open (PermissionError WinError 32).
         async with aiofiles.open(w_forecast_cache_path, "rb") as file:
             content = await file.read()
-            data = pickle.loads(content)
-            if not isinstance(data, pd.DataFrame) or data.empty:
-                self.logger.error("Cache file is corrupt or empty.")
-                self.logger.error(
-                    "Try running action `weather-forecast-cache` to pull new data from forecast API."
+        data = pickle.loads(content)
+        if not isinstance(data, pd.DataFrame) or data.empty:
+            self.logger.error("Cache file is corrupt or empty.")
+            self.logger.error(
+                "Try running action `weather-forecast-cache` to pull new data from forecast API."
+            )
+            try:
+                os.remove(w_forecast_cache_path)
+            except FileNotFoundError:
+                pass
+            return None
+        # Filter cached forecast data to match current forecast_dates start-end range
+        if self.forecast_dates[0] in data.index and self.forecast_dates[-1] in data.index:
+            data = data.loc[self.forecast_dates[0] : self.forecast_dates[-1]]
+            self.logger.info("Retrieved forecast data from the previously saved cache.")
+        else:
+            if self.weather_forecast_method == "open-meteo":
+                # Open-Meteo has no rate limits: delete the stale cache so
+                # the next call fetches fresh data directly from the API.
+                self.logger.warning(
+                    "Cache does not fully cover the requested timeframe. "
+                    "Removing stale Open-Meteo cache; fresh data will be fetched from the API."
                 )
                 try:
                     os.remove(w_forecast_cache_path)
                 except FileNotFoundError:
                     pass
                 return None
-            # Filter cached forecast data to match current forecast_dates start-end range
-            if self.forecast_dates[0] in data.index and self.forecast_dates[-1] in data.index:
-                data = data.loc[self.forecast_dates[0] : self.forecast_dates[-1]]
-                self.logger.info("Retrieved forecast data from the previously saved cache.")
             else:
-                if self.weather_forecast_method == "open-meteo":
-                    # Open-Meteo has no rate limits: delete the stale cache so
-                    # the next call fetches fresh data directly from the API.
-                    self.logger.warning(
-                        "Cache does not fully cover the requested timeframe. "
-                        "Removing stale Open-Meteo cache; fresh data will be fetched from the API."
-                    )
-                    try:
-                        os.remove(w_forecast_cache_path)
-                    except FileNotFoundError:
-                        pass
-                    return None
-                else:
-                    # Solcast and other rate-limited APIs: serve best-effort
-                    # stale data to avoid burning daily API quota.
-                    self.logger.warning(
-                        "Cache does not fully cover the requested timeframe. "
-                        "Serving best-effort stale data (reindexed + zero-filled). "
-                        "Cache preserved to protect API rate limits."
-                    )
-                    combined_index = data.index.union(self.forecast_dates).sort_values()
-                    data = data.reindex(combined_index)
-                    data.interpolate(method="time", inplace=True)
-                    data = data.reindex(self.forecast_dates)
+                # Solcast and other rate-limited APIs: serve best-effort
+                # stale data to avoid burning daily API quota.
+                self.logger.warning(
+                    "Cache does not fully cover the requested timeframe. "
+                    "Serving best-effort stale data (reindexed + zero-filled). "
+                    "Cache preserved to protect API rate limits."
+                )
+                combined_index = data.index.union(self.forecast_dates).sort_values()
+                data = data.reindex(combined_index)
+                data.interpolate(method="time", inplace=True)
+                data = data.reindex(self.forecast_dates)
 
-                    irradiance_cols = [c for c in ["ghi", "dni", "dhi"] if c in data.columns]
-                    other_cols = [c for c in data.columns if c not in irradiance_cols]
+                irradiance_cols = [c for c in ["ghi", "dni", "dhi"] if c in data.columns]
+                other_cols = [c for c in data.columns if c not in irradiance_cols]
 
-                    if other_cols:
-                        data[other_cols] = data[other_cols].ffill().bfill()
-                    if irradiance_cols:
-                        data[irradiance_cols] = data[irradiance_cols].fillna(0.0)
+                if other_cols:
+                    data[other_cols] = data[other_cols].ffill().bfill()
+                if irradiance_cols:
+                    data[irradiance_cols] = data[irradiance_cols].fillna(0.0)
 
-                    data = data.fillna(0.0)
-            return data
+                data = data.fillna(0.0)
+        return data
 
     async def set_cached_forecast_data(self, w_forecast_cache_path, data) -> pd.DataFrame:
         r"""

--- a/tests/test_forecast.py
+++ b/tests/test_forecast.py
@@ -1548,6 +1548,23 @@ class TestForecast(unittest.IsolatedAsyncioTestCase):
         self.assertIsNone(result)
         self.assertFalse(w_forecast_cache_path.exists(), "Stale Open-Meteo cache should be deleted")
 
+    async def test_get_cached_forecast_data_corrupt_cache_deletes_and_returns_none(self):
+        """A corrupt (non-DataFrame) cache pickle must be deleted and return None.
+
+        Also exercises the os.remove-after-close path on Windows (the file handle
+        must be released before unlink, else PermissionError [WinError 32]).
+        """
+        w_forecast_cache_path = emhass_conf["data_path"] / "weather_forecast_data_corrupt_test.pkl"
+        # Write a non-DataFrame payload directly (bypasses set_cached_forecast_data)
+        with open(w_forecast_cache_path, "wb") as f:
+            pickle.dump({"not": "a dataframe"}, f)
+        self.assertTrue(w_forecast_cache_path.exists())
+
+        result = await self.fcst.get_cached_forecast_data(w_forecast_cache_path)
+
+        self.assertIsNone(result)
+        self.assertFalse(w_forecast_cache_path.exists(), "Corrupt cache should be deleted")
+
     async def test_get_cached_forecast_data_stale_solcast_zero_fills(self):
         """Stale Solcast cache must be served as reindexed/zero-filled data.
 

